File: 백준/Silver/2512. 예산/예산.py
import sys

class Solution:
    def budget(self):
        input = sys.stdin.readline
        N = int(input())
        city = list(map(int, input().split()))
        total_budget = int(input())
        city.sort()

        start, end = 0, city[-1]
        while start <= end:
            mid = (start + end) // 2
            total = 0
            for i in city:
                if i <= mid:
                    total += i
                else:
                    total += mid
            if total <= total_budget:
                start = mid + 1
            else:
                end = mid - 1
        print(end)

Solution().budget()

File: 프로그래머스/1/12982. 예산/예산.java
import java.util.*;

class Solution {
    public int solution(int[] d, int budget) {
        int answer = 0;
        
        Arrays.sort(d); // 1,2,3,4,5
        for(int i = 0; i < d.length; i++) {
            int add = 0;
            int len = 0;
            for(int j = i; j < d.length; j++) {
                if(add + d[j] <= budget) {
                    add += d[j];
                    len++;
                } else {
                    break;
                }
            }
            if(answer < len) answer = len;
        }        
        
        return answer;
    }
}